docs/plugins.md

```diff
@@ -0,0 +1,174 @@
+# BrighterScript plugins
+
+The brighterscript compiler is extensible using **JavaScript** plugins.
+
+## Configuration
+
+### Declarative configuration
+
+You can refer to scripts / npm modules directly in the `bsconfig.json`.
+
+Those plugins will be loaded by the VSCode extension and can provide live diagnostics within the editor.
+
+```json
+{
+    "plugins": [
+        "./scripts/myPlugin.js",
+        "@rokucommunity/bslint"
+    ]
+}
+```
+
+### Programmatic configuration
+
+When using the compiler API directly, plugins can directly reference your code:
+
+```typescript
+import { ProgramBuilder } from 'brighterscript';
+import myPlugin from './myPlugin';
+
+const builder = new ProgramBuilder();
+builder.addPlugin(myPlugin);
+builder.run(/*...*/);
+```
+
+## Compiler API
+
+Objects in the brighterscript compiler implement a number of events that you can listen to. These events allow you to peek / modify the objects that the compiler manipulates.
+
+An object re-emits all the events of the child objects it owns.
+
+The top level object is the `ProgramBuilder` which runs the overall process: preparation, running the compiler (`Program`) and producing the output.
+
+`Program` is the compiler itself; it handles the creation of program objects (`Scope`, `BrsFile`, `XmlFile`) and overal validation.
+
+`Scope` is a logical group of files; for instance `source` files, or each component's individual scope (XML definition, related components and imported scripts). A file can be included in multiple scopes.
+
+`BrsFile` is a `.brs` or `.bs` file; its AST can be modified by a plugin.
+
+`XmlFile` is a component; it doesn't have a proper AST currently but it can be "patched".
+
+```typescript
+export interface CompilerPlugin {
+    getName?: () => string;
```

Comment:
@elsassph is there a reason this is a function rather than just a string?

Reply:
I've gone back and forth with it. Prefer a string?

Reply:
Yeah, I think I do prefer a string.

Reply:
Ok done